README.md
# node-cache-manager-fs
Node Cache Manager store for Filesystem
=======================================

The Filesystem store for the [node-cache-manager](https://github.com/BryanDonovan/node-cache-manager) module.

## Installation

```sh
npm install cache-manager-fs --save
```

## Features

* Saves anything that is `JSON.stringify`-able to disk
* limit maximum size on disk
* refill cache on startup (in case of application restart)
* "Callback"-interface style (no promises, no async)

## Usage example

Here are examples that demonstrate how to implement the Filesystem cache store.

## Single store

```javascript
// node cachemanager
var cacheManager = require('cache-manager');
// storage for the cachemanager
var fsStore = require('cache-manager-fs');
// initialize caching on disk
const diskCache = cacheManager.caching({
    store: fsStore, 
    options: {
        path: 'diskcache',          //path for cached files
        ttl: 60 * 60,               //time to life in seconds
        maxsize: 1000*1000*1000,    // max size in bytes on disk
        zip: true,                  //zip files to save diskspace (default: false)        
        preventfill:true            
    }
});

(async () => {

    await diskCache.set('key', 'value');
    console.log(await diskCache.get('key')); //"value"
    console.log(await diskCache.ttl('key')); //3600 seconds
    await diskCache.del('key');
    console.log(await diskCache.get('key')); //undefined

    console.log(await getUserCached(5)); //{id: 5, name: '...'}
    console.log(await getUserCached(5)); //{id: 5, name: '...'}

    await diskCache.reset();

    function getUserCached(userId) {
        return diskCache.wrap(userId /* cache key */, function () {
            return getUser(userId);
        });
    }

    async function getUser(userId) {
        return {id: userId, name: '...'};
    }

})();

```

### Options

options for store initialization

```javascript

options.ttl = 60; // time to life in seconds
options.path = 'cache/'; // path for cached files
options.preventfill = false; // prevent filling of the cache with the files from the cache-directory
options.fillcallback = null; // callback fired after the initial cache filling is completed
options.zip = false; // if true the cached files will be zipped to save diskspace
options.reviveBuffers = true; // if true buffers are returned from cache as buffers, not objects

```
## Installation

    npm install cache-manager-fs
	
## Tests

To run tests:

npm test

## Code Coverage

To run Coverage:

npm run coverage

## License

cache-manager-fs is licensed under the MIT license.


CHANGELOG.md
# Changelog

##  [1.0.12] - 2023-08-07

- add zip-support when filling the cache on startup
- work on unittests
- add CHANGELOG.md
- publish new version

##  [1.0.11] - 2023-08-07

- added open pullrequests 
- published new version


test/index.test.js
var assert = require('chai').assert;
var store = require('../index.js')
var fs = require('fs');
const removeDir = require('rimraf');
var path = require('path');
var uuid = require('uuid');
const cacheDirectory = __dirname + '/cache';

function countFilesInCacheDir() {
    return fs.readdirSync(cacheDirectory).length;
}

describe('tests for the disk-store module', function () {

	// remove test directory before each test
	beforeEach(function (done) {
		removeDir(cacheDirectory, function (err) {
			cache = store.create({path: cacheDirectory});
			done(err);
		});
	});
	// remove test directory after last test
	after(function (done) {
		removeDir(cacheDirectory, done);
	});

	describe('construction', function () {
        it('should create cache dir', function () {
            assert(fs.existsSync(cache.options.path));
        });

		it('simple create cache test', function ()
		{
			// create a store with default values
			var s = store.create();
			// check the creation result
			assert.isObject(s);
			assert.isObject(s.options);
			assert.isTrue(fs.existsSync(s.options.path));
		});

		it('create cache with option path test', function () {
			// create a store
			var s = store.create({options: {path:cacheDirectory, preventfill:true}});
			// check path option creation
			assert.isObject(s);
			assert.isObject(s.options);
			assert.isTrue(fs.existsSync(s.options.path));
			assert(s.options.path == cacheDirectory);
		});
	});

	describe('get', function () {

		it('should return undefined on non existing key callback', function (done)
		{
			var s=store.create({options: {path:cacheDirectory, preventfill:true}});
			s.get('asdf', function (err, data)
			{
                assert.strictEqual(null, err);
				assert.strictEqual(null, data);
//				assert(data === null);
				done();
			});
		});

		describe('test missing file on disk', function() {
			it('filename empty', function (done){
				var s=store.create({options: {path:cacheDirectory, preventfill:true}});
				s.set('test','test', function (err)
				{
					assert(err === null);
					var tmpfilename = s.collection['test'].filename;
					s.collection['test'].filename = null;
					s.get('test', function (err,data) {
						assert(err !== null);
						assert(data == null);
						s.collection['test'].filename = tmpfilename;
						s.del('test', function (err)
						{
							assert(err == null);
							done();
						});
					})
				});
			});

			it('file does not exist', function (done){
				var s=store.create({options: {path:cacheDirectory, preventfill:true}});
				s.set('test','test', function (err)
				{
					assert(err === null);
					var tmpfilename = s.collection['test'].filename;
					s.collection['test'].filename = "bla";
					s.get('test', function (err,data) {
						assert(err !== null);
						assert(data == null);
						s.collection['test'].filename = tmpfilename;
						s.del('test', function (err)
						{
							assert(err == null);
							done();
						});
					})
				});
			});
		});

		it('test expired of key (and also ttl option on setting)', function (done)
		{
			var s=store.create({options: {path:cacheDirectory, preventfill:true}});
			s.set('asdf','blabla', {ttl:-1000}, function (err)
			{
				assert(err === null)
				s.get('asdf',function (err,data){
					assert(err === null, 'error is not null!'+err);
					assert(data === null);
					done();
				})
			});
		})
	});

	describe('set', function () {

		it('simple set test', function (done)
		{
			var s=store.create({options: {path:cacheDirectory, preventfill:true}});
			var data = 'a lot of data in a file'
			s.set('asdf',data, function (err,data2)
			{
				assert(err === null);
				assert(data2,'check if entry has been returned on insert');
				s.get('asdf', function (err, data2)
				{
					assert(data2,'check if entry could be retrieved');
					assert(data === data2);
					done();
				});
			});
		});
	});

	describe('keys', function() {

		it('simple keys test', function (done) {
			var s=store.create({options: {path:cacheDirectory, preventfill:true}});
			var data = 'just a string with data';
			s.set('key123', data, function (err, data2) {
				assert(err === null);
				s.keys(function(err, keys) {
					assert(err === null);
					assert(keys.length === 1);
					assert(keys[0] === 'key123');
					done();
				});
			});
		});
	});

	describe('del / reset', function () {

		it('simple del test for not existing key', function (done)
		{
			var s=store.create({options: {path:cacheDirectory, preventfill:true}});
			s.del('not existing', function (err) {
				done();
			});
		});

		it('successfull deletion', function (done)
		{
			var s=store.create({options: {path:cacheDirectory, preventfill:true}});
			s.set('nix','empty', function (err) {
				assert(err === null);
				s.reset('nix', function (err) {
					done();
				});
			});
		});

		describe('delete errorhandling', function() {
			it('file not exists', function(done) {
				var s=store.create({options: {path:cacheDirectory, preventfill:true}});
				s.set('test','empty', function(err) {
					assert(err === null);
					var fn = s.collection['test'].filename;
					s.collection['test'].filename = s.collection['test'].filename+".not_here";
					s.del('test', function(err) {
						s.collection['test'].filename = fn;
						assert(err==null);
						done();
					});
				});
			});


			it('filename not set', function(done) {
				var s=store.create({options: {path:cacheDirectory, preventfill:true}});
				s.set('test','empty', function(err) {
					assert(err === null);
					var fn = s.collection['test'].filename;
					s.collection['test'].filename = null;
					s.del('test', function(err) {
						s.collection['test'].filename = fn;
						assert(err==null);
						done();
					});
				})
			});

		})

		it('reset all', function(done) {
			var s=store.create({options: {path:cacheDirectory, preventfill:true}});
			s.set('test', 'test', function(err){
				assert(err === null);

				s.set('test2', 'test2', function(err) {
					assert(err === null);
					s.reset(function(err) {
						assert(err === null);

						s.keys(function(err, keys) {
							assert(err === null);
							assert(keys.length === 0);
							done();
						});
					});
				});
			});
		});

		it('reset callback', function (done)
		{
			var s=store.create({options: {path:cacheDirectory, preventfill:true}});
			s.set('test','test', function (err)
			{
				assert(err === null);
				s.reset(function (error) {
					assert(err === null);
					done();
				})
			});
		});
	});

	describe('isCacheableValue', function () {

		it('works', function () {
			var s=store.create({options: {path:cacheDirectory, preventfill:true}});
			assert(!s.isCacheableValue(null));
			assert(!s.isCacheableValue(undefined));
		});
	});

    describe('zip test', function() {
       it('save and load again', function(done) {
			// create store
			var s=store.create({options: {zip:true, path:cacheDirectory, preventfill:true}});
            var datastring = "bla only for test \n and so on...";
            var dataKey = "zipDataTest";
			s.set(dataKey, datastring, function (err) {
				assert(err === null);
                s.get(dataKey, function (err, data) {
                    assert.strictEqual(err, null);
                    assert.strictEqual(data, datastring);
                    done();
                });
            });
       })
    });

	describe('integrationtests', function () {

		it('cache initialization on start', function (done) {
			// create store
			var s=store.create({options: {path:cacheDirectory, preventfill:true}});
			// save element
			s.set('RestoreDontSurvive', 'data', {ttl:-1}, function (err) {
				assert(err === null);
				s.set('RestoreTest','test', function (err)
				{
					var t=store.create({options: {path:cacheDirectory, fillcallback: function () {
						//fill complete
						t.get('RestoreTest', function (err, data) {
							assert(data === 'test');
							t.get('RestoreDontSurvive', function (err,data) {
								assert.strictEqual(err, null);
								assert.strictEqual(data, null);
								assert(s.currentsize > 0, 'current size not correctly initialized - '+s.currentsize);
								done();
							});
						});
					}
					}});
				});
			});
		});

		it('cache initialization with zip', function (done) {
			var s=store.create({options: {path:cacheDirectory, preventfill:true, zip:true}});
			s.set('testkey','value', function (err) { 
				var t = store.create({ options: {path:cacheDirectory, zip:true, fillcallback: function () {
					//fill complete
					t.get('testkey', function (err, data) {
						assert.equal('value',data);
						done();
					});
				}}});
			});
		});

		it('max size option', function (done) {
			// create store
			var s = store.create({
				options: {
					path: cacheDirectory,
					preventfill: true,
					maxsize: 1
				}
			});

			s.set('one', 'dataone', {}, function (err, val) {
				assert.strictEqual(err, 'Item size too big.');
				assert.strictEqual(Object.keys(s.collection).length, 0);

				s.set('x', 'x', { ttl: -1 }, function (err, val) {
					assert.strictEqual(err, 'Item size too big.');
					assert(Object.keys(s.collection).length === 0);

					s.options.maxsize = 300;
					s.set('a', 'a', { ttl: 10000 }, function (err, val) {
						assert.strictEqual(err, null);
						assert.equal(Object.keys(s.collection).length, 1);

						s.set('b', 'b', { ttl: 0 }, function (err){
							assert.strictEqual(err, null);

							s.set('c', 'c', { ttl: 100 }, function (err){
								assert.strictEqual(err, null);

								// now b should be removed from the cache, a should exists
								s.get('a', function (err, data) {
									assert.strictEqual(err, null);
									assert.equal(data,'a');

									s.get('b', function (err,data){
										assert.strictEqual(err, null);
										assert.strictEqual(data, null);
										done();
									});
								});
							});
						});
					});
				});
			});
		});

        describe('can handle unexpected files in the cache directory', function () {
            var originalUUIDv4;
            var mockUUID;
            var s;

            beforeEach(function () {
                originalUUIDv4 = uuid.v4;
                uuid.v4 = function mockv4() {
                    if (mockUUID) {
                        return mockUUID
                    };
                    return originalUUIDv4();
                }
                s = store.create({ options: { path:cacheDirectory, preventfill: true } });

                // simulate cruft files that may exist in the cache directory
                fs.writeFileSync(path.join(cacheDirectory, '.DS_Store'), 'not JSON data');
            });

            afterEach(function () {
                s.cleancache();
            });

            after(function () {
                uuid.v4 = originalUUIDv4;
            })

            it('skips over non-cache files', function (done) {
                // create store and set some entries
                s.set('key0', 'data0', function (err) {

                    // ungracefully reset the cache
                    s.collection = {};
                    s.currentsize = 0;

                    s.intializefill(function (err) {
                        s.get('key0', function (err, data) {
                            assert.equal(data, 'data0', `expected "data0", but received "${data}"`);
                            fs.readdir(cacheDirectory, function (err, files) {
                                assert.isTrue(files.includes('.DS_Store'), '".DS_Store" should exist.');
                                done();
                            });
                        });
                    });
                });
            });

            it('truncated JSON files', function (done) {
                mockUUID = '8f1a33a9-9984-4e98-9c72-3bea37dab031';
                var cacheFileName = 'cache_' + mockUUID + '.dat';
                var cachedFilePath = path.join(cacheDirectory, cacheFileName);

                // set an entry
                s.set('getTruncated', 'some data...', function (err) {
                    // simulate truncated .dat file
                    var tmpData = fs.readFileSync(cachedFilePath, 'utf8');
                    fs.writeFileSync(cachedFilePath, tmpData.substring(0, 20), 'utf8');

                    // ungracefully reset the cache
                    s.collection = {};
                    s.currentsize = 0;

                    // re-init the cache from files saved in directory
                    s.intializefill(function () {
                        // Callback to run after fill has completed
                        s.get('getTruncated', function (err, data) {
                            var files = fs.readdirSync(cacheDirectory);
                            assert.equal(data, null, `expected null but received "${data}"`);
                            assert.isFalse(files.includes(cacheFileName), `"${cacheFileName}" should not exist`);
                            done();
                        });
                    });
                });
            });
        });
	});
});


index.js
'use strict';

/**
 * Module dependencies
 */
var noop = function () {};
var fs = require("fs");
var fsp = require("fs-promise");
var crypto = require('crypto');
var path = require('path');
var async = require('async');
var extend = require('extend');
var uuid = require('uuid');
var zlib = require('zlib');
const gzip = zlib.createGzip();

/**
 * Export 'DiskStore'
 */

module.exports = {
  create : function (args) {
		return new DiskStore(args && args.options ? args.options : args);
  }
};

/**
 * Helper function that revives buffers from object representation on JSON.parse
 */
function bufferReviver(k, v) {
	if (
		v !== null &&
		typeof v === 'object' &&
		'type' in v &&
		v.type === 'Buffer' &&
		'data' in v &&
		Array.isArray(v.data)) {
		return new Buffer(v.data);
	}
	return v;
}

/**
 * helper object with meta-informations about the cached data
 */
function MetaData () {

	// the key for the storing
	this.key = null;
	// data to store
	this.value = null;
	// temporary filename for the cached file because filenames cannot represend urls completely
	this.filename = null;
	// expirydate of the entry
	this.expires = null;
	// size of the current entry
	this.size = null;
}

/**
 * construction of the disk storage
 */
function DiskStore (options) {
	options = options || {};

	this.options = extend({
		path: 'cache/',
		ttl: 60,
		maxsize: 0,
        zip: false
	}, options);

	// check storage directory for existence (or create it)
	if (!fs.existsSync(this.options.path)) {
			fs.mkdirSync(this.options.path);
	}

	this.name = 'diskstore';

	// current size of the cache
	this.currentsize = 0;

	// internal array for informations about the cached files - resists in memory
	this.collection = {};

	// fill the cache on startup with already existing files
	if (!options.preventfill) {
		this.intializefill(options.fillcallback);
	}
}

/**
 * indicate, whether a key is cacheable
 */
DiskStore.prototype.isCacheableValue = function (value) {

	return value !== null && value !== undefined;
};

/**
 * delete an entry from the cache
 */
DiskStore.prototype.del = function (key, options, cb) {
  
  if (typeof options === 'function') {
        cb = options;
        options = null;
  }
  cb = typeof cb === 'function' ? cb : noop;

  // get the metainformations for the key
	var metaData = this.collection[key];
	if (!metaData) {
		return cb(null);
	}

  // check if the filename is set
  if (!metaData.filename) {
	  return cb(null);
  }
  // check for existance of the file
  fsp.exists(metaData.filename).
  then(function(exists) {
	if (exists) {
		return;
	}
	reject();
  })
  .then(function() {
	  // delete the file
	  return fsp.unlink(metaData.filename);
  }, function() {
	  // not found
	  cb(null);
  }).then(function() {
	  // update internal properties
	  this.currentsize -= metaData.size;
	  this.collection[key] = null;
	  delete this.collection[key];
      cb(null);
  }.bind(this)).catch(function(err) {
	  cb(null);
  });
};


/**
 * zip an input string if options want that
 */
DiskStore.prototype.zipIfNeeded = function(data, cb)
{
    if (this.options.zip)
    {
        zlib.deflate(data, function(err, buffer) {
        if (!err) {
            cb(null, buffer);
        }
        else
        {
            cb(err, null);
        }
        });
    }
    else
    {
        cb(null, data);
    }
}

/**
 *unpzip an input string if options want that
 */
DiskStore.prototype.unzipIfNeeded = function (data, cb) {
    if (this.options.zip) {
        zlib.unzip(data, function (err, buffer) {
            if (!err) {
                cb(null, buffer);
            }
            else {
                cb(err, null);
            }
        });
    }
    else {
        cb(null, data);
    }
}

/**
 * set a key into the cache
 */
DiskStore.prototype.set = function (key, val, options, cb) {

	cb = typeof cb === 'function' ? cb : noop;

  if (typeof options === 'function') {
		cb = options;
		options = null;
  }

  // get ttl
  var ttl = (options && (options.ttl || options.ttl === 0)) ? options.ttl : this.options.ttl;

  var metaData = extend({}, new MetaData(), {
  	key: key,
  	value: val,
  	expires: Date.now() + ((ttl || 60) * 1000),
  	filename: this.options.path + '/cache_' + uuid.v4() + '.dat'
  });

  var stream = JSON.stringify(metaData);

  metaData.size = stream.length;

  if (this.options.maxsize && metaData.size > this.options.maxsize) {
  	return cb('Item size too big.');
  }


  // remove the key from the cache (if it already existed, this updates also the current size of the store)
  this.del(key, function (err) {

		if (err) {
	  	return cb(err);
		}

		// check used space and remove entries if we use to much space
		this.freeupspace(function () {

		  try {
                this.zipIfNeeded(stream, function(err, processedStream) {

                    // write data into the cache-file
                    fs.writeFile(metaData.filename, processedStream, function (err) {

                        if (err) {
                                return cb(err);
                        }

                        // remove data value from memory
                        metaData.value = null;
                        delete metaData.value;

                        this.currentsize += metaData.size;

                        // place element with metainfos in internal collection
                        this.collection[metaData.key] = metaData;
                        return cb(null, val);

                    }.bind(this));
                }.bind(this));

		  } catch(err) {

				return cb(err);
		  }

		}.bind(this));

  }.bind(this));

};

/**
 * helper method to free up space in the cache (regarding the given spacelimit)
 */
DiskStore.prototype.freeupspace = function (cb) {

  cb = typeof cb === 'function' ? cb : noop;

  if (!this.options.maxsize) {
		return cb(null);
  }

  // do we use to much space? then cleanup first the expired elements
  if (this.currentsize > this.options.maxsize) {
		this.cleanExpired();
  }

  // when the spaceusage is to high, remove the oldest entries until we gain enough diskspace
  if (this.currentsize <= this.options.maxsize) {
  	return cb(null);
  }

	// for this we need a sorted list basend on the expire date of the entries (descending)
	var tuples = [], key;
	for (key in this.collection) {
		tuples.push([key, this.collection[key].expires]);
	}

	tuples.sort(function sort (a, b) {

		a = a[1];
		b = b[1];
		return a < b ? 1 : (a > b ? -1 : 0);
	});

	return this.freeupspacehelper(tuples, cb);
};

/**
 * freeup helper for asnyc space freeup
 */
DiskStore.prototype.freeupspacehelper = function (tuples, cb) {

	// check, if we have any entry to process
	if (tuples.length === 0) {
		return cb(null);
	}

	// get an entry from the list
	var tuple = tuples.pop();
	var key = tuple[0];

	// delete an entry from the store
	this.del(key, function deleted (err) {

		// return when an error occures
		if (err) {
		  return cb(err);
		}

		// stop processing when enouth space has been cleaned up
		if (this.currentsize <= this.options.maxsize) {
			return cb(err);
		}

		// ok - we need to free up more space
		return this.freeupspacehelper(tuples, cb);
	}.bind(this));
};

/**
 * get entry from the cache
 */
DiskStore.prototype.get = function (key, options, cb) {

	if (typeof options === 'function') {
		cb = options;
	}

	cb = typeof cb === 'function' ? cb : noop;

  // get the metadata from the collection
  var data = this.collection[key];

  if (!data) {

	  // not found
	  return cb(null, null);
  }

  // found but expired
  if (data.expires < new Date()) {

	  // delete the elemente from the store
	  this.del(key, function (err) {
		return cb(err, null);
	  });
  } else {

		// try to read the file
		try {

			fs.readFile(data.filename, function (err, fileContent) {
				if (err) {
					return cb(err);
				}

				this.unzipIfNeeded(fileContent, function(err, decompressedContent) {
					if (err) {
						return cb(err);
					}

					var diskdata;
					if(this.options.reviveBuffers) {
						diskdata = JSON.parse(decompressedContent, bufferReviver);
					} else {
						diskdata = JSON.parse(decompressedContent);
					}
					cb(null, diskdata.value);
				}.bind(this));
			}.bind(this));
		} catch(err) {
			cb(err);
		}
  	}
};

/**
 * get keys stored in cache
 * @param {Function} cb
 */
DiskStore.prototype.keys = function (cb) {

	cb = typeof cb === 'function' ? cb : noop;

	var keys = Object.keys(this.collection);

	cb(null, keys);
};

/**
 * cleanup cache on disk -> delete all used files from the cache
 */
DiskStore.prototype.reset = function (key, cb) {

  cb = typeof cb === 'function' ? cb : noop;

  if (typeof key === 'function') {
		cb = key;
		key = null;
  }

  if (Object.keys(this.collection).length === 0) {
		return cb(null);
  }

  try {

		// delete special key
		if (key !== null) {

		  this.del(key);
		  return cb(null);
		}

		async.eachSeries(this.collection,
			function (elementKey, callback) {
				this.del(elementKey.key, callback);
			}.bind(this),
			function (err) {
				cb(null);
			}
		);

  } catch(err) {

		return cb(err);
  }

};

/**
 * helper method to clean all expired files
 */
DiskStore.prototype.cleanExpired = function () {

	var key, entry;

  for (key in this.collection) {

		entry = this.collection[key];

		if (entry.expires < new Date()) {

		  this.del(entry.key);
		}
  }
}

/**
 * clean the complete cache and all(!) files in the cache directory
 */
DiskStore.prototype.cleancache = function (cb) {

	cb = typeof cb === 'function' ? cb : noop;

  // clean all current used files
  this.reset();

  // check, if other files still resist in the cache and clean them, too
  var files = fs.readdirSync(this.options.path);

  files
  	.map(function (file) {

	  	return path.join(this.options.path, file);
  	}.bind(this))
  	.filter(function (file) {

	  	return fs.statSync(file).isFile();
  	}.bind(this))
  	.forEach(function (file) {

	  	fs.unlinkSync(file);
  	}.bind(this));

  cb(null);

};

/**
 * fill the cache from the cache directory (usefull e.g. on server/service restart)
 */
DiskStore.prototype.intializefill = function (cb) {

	cb = typeof cb === 'function' ? cb : noop;

  // get the current working directory
  fs.readdir(this.options.path, function (err, files) {

		// get potential files from disk
		files = files.map(function (filename) {

				return path.join(this.options.path, filename);
			}.bind(this)).filter(function (filename) {

				return fs.statSync(filename).isFile();
			}).filter(function (filename) {

				var re = /^cache_[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}.dat$/i
				return re.test(path.basename(filename));
			});

		// use async to process the files and send a callback after completion
		async.eachSeries(files, function (filename, callback) {

		  fs.readFile(filename, function (err, data) {

				// stop file processing when there was an reading error
				if (err) {
				  return callback();
				}

				this.unzipIfNeeded(data, function(err, unzippedData) {
					try {

						// get the json out of the data
						var diskdata = JSON.parse(unzippedData);
	  
					  } catch(err) {
	  
						// when the deserialize doesn't work, probably the file is uncomplete - so we delete it and ignore the error
						try {
							fs.unlinkSync(filename);
						} catch(ignore) {
	  
						}
	  
						return callback();
					  }
	  
					  // update the size in the metadata - this value isn't correctly stored in the file
					  diskdata.size = data.length;
	  
					  // update collection size
					  this.currentsize+=data.length;
	  
					  // remove the entrys content - we don't want the content in the memory (only the meta informations)
					  diskdata.value = null;
					  delete diskdata.value;
	  
					  // and put the entry in the store
					  this.collection[diskdata.key] = diskdata;
	  
					  // check for expiry - in this case we instantly delete the entry
					  if (diskdata.expires < new Date()) {
	  
						this.del(diskdata.key, function () {
	  
							  return callback();
						});
					  } else {
	  
						return callback();
					  }
				}.bind(this));
		 	}.bind(this));
		}.bind(this), function (err) {
			cb(err || null);
		});
  }.bind(this));

};
